File: backend/app/schemas/chatflow.py
"""Chatflow 流程定义、版本与调试运行的数据契约。"""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class ChatflowCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=2000)
    graph: dict | None = None


class ChatflowUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=2000)
    graph: dict | None = None
    enabled: bool | None = None


class ChatflowVersionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    chatflow_id: uuid.UUID
    version: int
    graph: dict
    note: str | None
    created_at: datetime


class ChatflowOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    description: str | None
    draft_graph: dict
    published_graph: dict
    published_version: int
    enabled: bool
    bound_assistant_count: int = 0
    created_at: datetime
    updated_at: datetime


class ChatflowListResponse(BaseModel):
    items: list[ChatflowOut]
    total: int


class ChatflowPublishRequest(BaseModel):
    note: str | None = Field(default=None, max_length=2000)


class ChatflowRollbackRequest(BaseModel):
    version: int = Field(ge=1)


class ChatflowDebugRequest(BaseModel):
    question: str = Field(min_length=1, max_length=1000)
    knowledge_base_id: uuid.UUID | None = None
    include_generation: bool = False


class ChatflowDebugResponse(BaseModel):
    nodes: list[dict]
    total_ms: float
    final: dict | None = None


class NodeTypeOut(BaseModel):
    type: str
    label: str
    category: str
    config_fields: list[dict] = Field(default_factory=list)


class NodeTypeListResponse(BaseModel):
    items: list[NodeTypeOut]
    total: int

File: frontend/src/types/chatflow.ts
export interface ChatflowNode {
  id: string
  type: string
  name: string
  enabled: boolean
  config: Record<string, unknown>
  next: string | null
}

export interface ChatflowGraph {
  nodes: ChatflowNode[]
  start_node_id: string
}

export interface ChatflowRecord {
  id: string
  name: string
  description: string | null
  draft_graph: ChatflowGraph
  published_graph: ChatflowGraph
  published_version: number
  enabled: boolean
  bound_assistant_count: number
  created_at: string
  updated_at: string
}

export interface ChatflowVersion {
  id: string
  chatflow_id: string
  version: number
  graph: ChatflowGraph
  note: string | null
  created_at: string
}

export interface NodeConfigField {
  key: string
  type: string
  label: string
}

export interface NodeTypeDef {
  type: string
  label: string
  category: string
  config_fields: NodeConfigField[]
}

export interface ChatflowDebugNode {
  id: string
  type: string
  name: string
  status: 'succeeded' | 'skipped' | 'failed'
  duration_ms: number
  output: Record<string, unknown>
  error: string | null
}

export interface ChatflowDebugResult {
  nodes: ChatflowDebugNode[]
  total_ms: number
  final: Record<string, unknown> | null
}
